src/schedule/smard_api.py

```diff
@@ -0,0 +1,295 @@
+
+            :param data: The dict
+            :return: A SmardApiIndex
+            """
+            return SmardApiIndex(**data)
+
+    timestamp = DateTimeField(unique=True)
+
+
+class SmardApiEntry(BaseModel):
+    """This class models an entry of the smard api. Each entry is linked to an
+    index timestamp. It contains the timestamp and the value of the entry.
+    The value is the amount of electricity produced from coal in
+    a quarter hour in megawatt hours.
+    """
+
+    class Schema(BaseModel.Schema):
+        """marshmallow schema for SmardApiEntry"""
+
+        timestamp = marsh.fields.DateTime(required=True)
+        value = marsh.fields.Float(required=True)
+        index_id = marsh.fields.UUID(required=True)
+
+        def _make(self, data: dict) -> "SmardApiEntry":
+            """Constructs a SmardApiEntry from a dict
+
+            :param data: The dict
+            :return: A SmardApiEntry
+            """
+            return SmardApiEntry(**data)
+
+    timestamp = DateTimeField(unique=True)
+    value = FloatField(null=True)
+    index_id = ForeignKeyField(SmardApiIndex, backref="entries")
+
+
+@dataclass
+class SmardDataAvailability:
+    """This class represents the availability of data from the smard api.
+    It is the return type of SmardApi.data_availability. If available is
+    True there is data available for the given time interval. If interval_altered
+    is True the time interval was altered because only a subset of the requested
+    interval contained data. The start and end attributes contain the actual
+    time interval for which data is available. They can be slitely different
+    from the requested time interval even if interval_altered is False to
+    match the exact time intervals of the data batches.
+    """
+
+    @classmethod
+    def none(cls):
+        """Returns a SmardDataAvailability object that represents the
+        case that no data is available.
+
+        :return: A SmardDataAvailability object
+        """
+        return cls(False, False, None, None)
+
+    available: bool
+    interval_altered: bool
+    start: datetime
+    end: datetime
+
+
+class SmardApi:
+    """This class provides an interface to the smard api. It provides
+    quarter hourly data of the amount of electricity produced from coal.
+    """
+
+    RESOLUTION: str = "quarterhour"
+    FILTER: str = "1223"  # Braunkohle
+    REGION: str = "50Hertz"  # name of the electricity grid company in eastern Germany
+    BASE_URL: str = f"https://www.smard.de/app/chart_data/{FILTER}/{REGION}"
+    MILLISECONDS_PER_SECOND: int = 1000
+
+    def __init__(self):
+        """Constructs a SmardApi"""
+        self._update_indices()
+
+    def _timestamp_to_dt(self, timestamp: int) -> datetime:
+        """Converts a timestamp to a datetime object
+
+        :param timestamp: The timestamp
+        :return: The datetime object
+        """
+        return datetime.fromtimestamp(timestamp / self.MILLISECONDS_PER_SECOND)
```

Comment:
I would suggest that you make the seconds file-scoped, so that you don't have this `self.---` which is kind of ugly in my opinion (same below). I don't know if the other constants should be file-scope as well.

Reply:
:+1: